src/util/result.h

```diff
@@ -6,37 +6,117 @@
 #define BITCOIN_UTIL_RESULT_H
 
 #include <util/translation.h>
-#include <variant>
+#include <optional>
 
-/*
- * 'BResult' is a generic class useful for wrapping a return object
- * (in case of success) or propagating the error cause.
-*/
-template<class T>
-class BResult {
-private:
-    std::variant<bilingual_str, T> m_variant;
+namespace util {
+
+//! Function result type intended for high-level functions that return error and
+//! warning strings in addition to normal result types.
+//!
+//! The Result<T> class is meant to be a drop-in replacement for
+//! std::optional<T> except it has additional methods to return error and
+//! warning strings for error reporting. Also, unlike std::optional, in order to
+//! support error handling in cases where callees need to pass additional
+//! information about failures back to callers, Result<T> objects can always be
+//! dereferenced regardless of the function's success or failure.
+//!
+//! This class is not intended to be used by low-level functions that do not
+//! return error or warning strings. These functions should use plain
+//! std::optional or std::variant types instead.
+//!
+//! See unit tests in result_tests.cpp for example usages.
+template<typename T>
+class Result;
+
+//! Tag types for result constructors.
+struct Error{};
+struct ErrorChain{};
+
+//! Result<void> specialization. Only returns errors and warnings, no values.
+template<>
+class Result<void>
+{
+protected:
+    std::vector<bilingual_str> m_errors;
+    std::vector<bilingual_str> m_warnings;
 
 public:
-    BResult() : m_variant(Untranslated("")) {}
-    BResult(const T& _obj) : m_variant(_obj) {}
-    BResult(const bilingual_str& error) : m_variant(error) {}
+    //! Success case constructor.
+    Result() = default;
 
-    /* Whether the function succeeded or not */
-    bool HasRes() const { return std::holds_alternative<T>(m_variant); }
+    //! Error case constructor for a single error.
+    Result(Error, bilingual_str error)
+    {
+        m_errors.emplace_back(std::move(error));
+    }
 
-    /* In case of success, the result object */
-    const T& GetObj() const {
-        assert(HasRes());
-        return std::get<T>(m_variant);
+    //! Error case constructor for a chained error.
+    Result(ErrorChain, bilingual_str error, Result<void>&& previous) : m_errors{std::move(previous.m_errors)}, m_warnings{std::move(previous.m_warnings)}
+    {
+        m_errors.emplace_back(std::move(error));
     }
 
-    /* In case of failure, the error cause */
-    const bilingual_str& GetError() const {
-        assert(!HasRes());
-        return std::get<bilingual_str>(m_variant);
+    //! Success check.
+    operator bool() const { return m_errors.empty(); }
+
+    //! Error retrieval.
+    const std::vector<bilingual_str>& GetErrors() const { return m_errors; }
+    std::tuple<const std::vector<bilingual_str>&, const std::vector<bilingual_str>&> GetErrorsAndWarnings() const { return {m_errors, m_warnings}; }
```

Comment:
Maybe a `GetWarnings()` method ?
I think there are some success cases that also return warnings.

Reply:
> Maybe a `GetWarnings()` method ? Or `m_errors` and `m_warnings` can be public. I think there are some success cases that also return warnings.

Yes sorry interface is just incomplete while in draft.